### DISCUSSION TITLE
Robust Flutter App Initialization Flow with Riverpod?

### DISCUSSION THREAD

**Discussion by bizz84** [rrousselGit/riverpod#3284] [Q&A]
### Some Context

Given this code:
```dart
void main() async {
  await someAsyncCodeThatMayThrow();
  runApp(const MyApp());
}
```
It is known that if `someAsyncCodeThatMayThrow()` throws, `runApp` will not execute and the app will remain stuck on the splash screen.

A partially better solution is this:

```dart
void main() async {
  try {
    await someAsyncCodeThatMayThrow();
    runApp(const MyApp());
  } catch (e, st) {
		log(e.toString(), stackTrace: st);
    runApp(const AppStartupErrorWidget(e));
  }
}
```

But it still has problems:
- no loading UI
- can't retry

### Improved solution

I've been thinking of a solution that allows me to:
- Asynchronously initialize dependencies
- Show some loading UI
- Handle errors and retry if something goes wrong

Here's an example I came up with:

```dart
@Riverpod(keepAlive: true)
Future<SharedPreferences> sharedPreferences(SharedPreferencesRef ref) =>
    SharedPreferences.getInstance();
    
@Riverpod(keepAlive: true)
Future<void> appStartup(AppStartupRef ref) async {
  // all asynchronous app initialization code should belong here:
  await ref.watch(sharedPreferencesProvider.future);
}

/// Widget class to manage asynchronous app initialization
class AppStartupWidget extends ConsumerWidget {
  const AppStartupWidget({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    // eagerly initialize appStartupProvider and all the providers it depends on
    final appStartupState = ref.watch(appStartupProvider);
    return appStartupState.when(
      data: (_) => MyApp(),
      loading: () => const AppStartupLoadingWidget(),
      error: (e, st) => AppStartupErrorWidget(
        message: e.toString(),
				onRetry: () {
          ref.invalidate(sharedPreferencesProvider);
          ref.invalidate(appStartupProvider);
        },      
      ),
    );
  }
}

void main() {
  // use it like this:
  runApp(const AppStartupWidget());
}
```

This setup gives us everything I need

- As soon as the app starts, the `AppStartupWidget` is loaded
- This causes the `appStartupProvider` to be loaded (along with all the providers it depends on)
- While the provider is loading, show a custom `AppStartupLoadingWidget`
- If there’s an error, show an `AppStartupErrorWidget` with a retry option
- If the retry option is selected, invalidate all the providers that need to be reloaded

On the other hand, if the initialization is successful, then the `MyApp` widget will be returned.

And from that point onwards, we can access the dependency like this:

```dart
@override
  Widget build(BuildContext context, WidgetRef ref) {
    final sharedPreferences = ref.watch(sharedPreferencesProvider).requireValue;
  }
```  
### Additional considerations

Now, suppose we have multiple providers, and other async initialization code to execute.

Would it make sense to do this?

```dart
@Riverpod(keepAlive: true)
Future<void> appStartup(AppStartupRef ref) async {
  await Future.wait([
    // Firebase init
    Firebase.initializeApp(options: DefaultFirebaseOptions.currentPlatform),
    // list of providers to be warmed up
    ref.watch(onboardingRepositoryProvider.future),
    ref.watch(someOtherProvider.future),
    // and so on
  ]);
}
```

One possible issue I see here is that if there's an error, it's hard to know which provider caused it.
So would it make sense to invalidate them all?

```dart
onRetry: () {
          ref.invalidate(someOtherProvider);
          ref.invalidate(sharedPreferencesProvider);
          ref.invalidate(appStartupProvider);
        }
```        

Are there any best practices in this regards?

**Comment by rrousselGit** [comment]
> One possible issue I see here is that if there's an error, it's hard to know which provider caused it.
So would it make sense to invalidate them all?

You likely want providers to retry themselves instead.
There's no reason to have the retry logic out of the provider and in the UI.
